### PR TITLE
Watchdog concurrency fix

### PR DESCRIPTION
**Pull request by janet-harmony**
* Add performance options to config
* Worker pool for http request, manager per rpc request
* WaitGroup to block multiple read & writes on Snapshot

### CODE REVIEW

File: monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go [hunk 1]
```diff
@@ -202,21 +200,31 @@ func (m *monitor) renderReport(w http.ResponseWriter, req *http.Request) {
...
-		DownMachineCount: linq.From(m.NoReplyMachines).Select(
+		NoReply:    totalNoReplyMachines,
+		DownMachineCount: linq.From(totalNoReplyMachines).Select(
 			func(c interface{}) interface{} { return c.(noReply).IP },
 		).Distinct().Count(),
 	})
-	m.SummarySnapshot = sum
+	m.use()
+	m.SummarySnapshot = make(map[string]map[string]interface{})
+	copier.Copy(&m.SummarySnapshot, &sum)
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: I don't want to use this library, it relies on reflection https://github.com/jinzhu/copier/blob/master/copier.go#L14 , that is very slow. Hit to performance not worth the convenience

File: monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go [hunk 2]
```diff
@@ -294,91 +315,162 @@ type headerInfoRPCResult struct {
...
-		rpcPayload []byte
-		rpcResult  []byte
-		oops       error
+func (m* monitor) manager(jobs chan work, interval int, nodeList []string,
+					rpc string, group *sync.WaitGroup, channels map[string](chan reply)) {
+	for now := range time.Tick(time.Duration(interval) * time.Second) {
+		queryID := 0
+		group.Add(len(nodeList))
+		for _, n := range nodeList {
+			requestBody, _ := json.Marshal(map[string]interface{} {
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: don't want to make majority of this data over & over. Make it once for each RPC, then just update the key "id"'s value

File: monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go [hunk 3]
```diff
@@ -294,91 +315,162 @@ type headerInfoRPCResult struct {
...
+		first := true
+		switch rpc {
+		case metadataRPC:
+			for d := range channels[rpc] {
+				if first {
+					m.WorkingMetadata.Down = []noReply{}
+					m.WorkingMetadata.Nodes = []metadataRPCResult{}
+					first = false
+				}
+				if d.oops != nil {
```
Comment (fxfactorial) [inline monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: invert the switch with the range, majority of the case code is the same

Comment (janet-harmony) [reply monitoring/harmony-monitor/blockchain-watchdog/reporting-server.go]: The only part of the code that can be shared is the range.

**Comment by fxfactorial** [comment]
great work!

**Pull request merged**